File: 0085-maximal-rectangle/0085-maximal-rectangle.cpp
class Solution {
public:
    int calcArea(vector<int>&h){
        int n=h.size();
        stack<int>s1,s2;
        vector<int>l(n),r(n);
        for(int i=0;i<n;i++){
            while(s1.size()&& h[s1.top()]>=h[i])s1.pop();
            l[i]=(s1.size())?s1.top():-1;
            s1.push(i);
        }
        for(int i=n-1;i>=0;i--){
            while(s2.size()&& h[s2.top()]>=h[i])s2.pop();
            r[i]=(s2.size())?s2.top():n;
            s2.push(i);
        }
        int ans=0;
        for(int i=0;i<n;i++){
            int left=i-l[i];
            int right=r[i]-i;
            int area=(left+right-1)*h[i];
            ans=max(ans,area);
        }
        return ans;
    }
    int maximalRectangle(vector<vector<char>>& matrix) {
        int n=matrix.size(),m=matrix[0].size();
        vector<int>vec(m,0);
        int ans=0;
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(matrix[i][j]=='0')vec[j]=0;
                else vec[j]++;
            }
            ans=max(ans,calcArea(vec));
        }
        return ans;
    }
};

File: 0085-maximal-rectangle/0085-maximal-rectangle.java
class Solution {
     public int func(int[] heights) {
        Stack<Integer> dp = new Stack<>();
        dp.push(0);
        int n = heights.length;
        int ans = 0;
        for (int i = 1; i <= n; i++) {
            while (!dp.isEmpty()) {
                if (i != n && heights[i] == heights[dp.peek()]) {
                    break;
                }
                if (i != n && heights[i] > heights[dp.peek()]) {
                    dp.push(i);
                    break;
                } else {
                    int ci = dp.pop();
                    int x;
                    if (dp.isEmpty()) x = i;
                    else x = i - dp.peek() - 1;
                    ans = Math.max(ans, x * heights[ci]);
                }
            }
            dp.push(i);
        }
        return ans;
    }

    public int maximalRectangle(char[][] matrix) {
        int ans = 0;
        int n = matrix.length;
        int m = matrix[0].length;
        int[][] mat = new int[n][m];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (matrix[i][j] == '1') mat[i][j] = 1;
            }
        }
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (i >= 1 && mat[i][j] == 1) mat[i][j] += mat[i - 1][j];
            }
            ans = Math.max(ans, func(mat[i]));
        }
        return ans;
    }
}